File: javalibrary/src/main/java/com/bluthlee/javalibrary/ui/RecorderWidgetManager.java
package com.bluthlee.javalibrary.ui;

import android.content.Context;
import android.view.MotionEvent;
import android.view.View;

import com.bluthlee.javalibrary.data.DataBinder;
import com.bluthlee.javalibrary.ui.recorderWidget.DragRecorderWidget;
import com.bluthlee.javalibrary.ui.recorderWidget.FloatRecorderWidget;
import com.bluthlee.javalibrary.ui.recorderWidget.RecorderWidget;

/**
 * Created by LC on 2017/8/31.
 */

public class RecorderWidgetManager {

    public static RecorderWidget initDragAudioRecorder(View recordButton, Context context, DataBinder dataBinder) {
        DragRecorderWidget dragRecorderWidget = (DragRecorderWidget) RecorderWidgetFactory.getRecorderWidget(DataBinder.STYLE_DRAG, context, dataBinder);
        setTouchListener(recordButton, dragRecorderWidget);
        return dragRecorderWidget;
    }

    public static RecorderWidget initFloatAudioRecorder(Context context, DataBinder dataBinder) {
        FloatRecorderWidget floatRecorderWidget = (FloatRecorderWidget) RecorderWidgetFactory.getRecorderWidget(DataBinder.STYLE_FLOAT, context, dataBinder);
        floatRecorderWidget.setStatus(RecorderWidget.STATUS_IDLE);
        setTouchListener(floatRecorderWidget.recorderView, floatRecorderWidget);
        return floatRecorderWidget;
    }

    private static void setTouchListener(final View recordButton, final RecorderWidget recorderWidget) {
        final boolean[] isCanceled = {false};
        final float[] downY = {0};
        recordButton.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View view, MotionEvent motionEvent) {
                switch (motionEvent.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        //开始录音
                        recorderWidget.setStatus(RecorderWidget.STATUS_START);
                        downY[0] = motionEvent.getY();
                        return true;
                    case MotionEvent.ACTION_MOVE:
                        float moveY = motionEvent.getY();
                        if (downY[0] - moveY > 100) {
                            recorderWidget.setStatus(RecorderWidget.STATUS_DRAG);
                            isCanceled[0] = true;
                        } else {
                            recorderWidget.setStatus(RecorderWidget.STATUS_RESUME);
                            isCanceled[0] = false;
                        }
                        return true;
                    case MotionEvent.ACTION_UP:
                        //停止录音
                        recordButton.performClick();
                        if (isCanceled[0]) {
                            //取消录音
                            recorderWidget.setStatus(RecorderWidget.STATUS_CANCEL);
                            isCanceled[0] = false;
                        } else {
                            //完成录音
                            recorderWidget.setStatus(RecorderWidget.STATUS_FINISH);
                        }
                        return true;
                }
                return false;
            }
        });
    }

}

File: kotlinlibrary/src/main/java/com/bluthlee/library/ui/RecorderWidgetManager.kt
package com.bluthlee.library.ui

import android.content.Context
import android.view.MotionEvent
import android.view.View
import com.bluthlee.library.data.DataBinder
import com.bluthlee.library.ui.recorderWidget.DragRecorderWidget
import com.bluthlee.library.ui.recorderWidget.FloatRecorderWidget
import com.bluthlee.library.ui.recorderWidget.RecorderWidget

/**
 * Created by LC on 2017/8/24.
 */
internal class RecorderWidgetManager {

    companion object {

        fun initDragAudioRecorder(recordButton: View, context: Context, dataBinder: DataBinder): RecorderWidget {
            val recorderWidget = RecorderWidgetFactory.getRecorderWidget(DataBinder.STYLE_DRAG, context, dataBinder) as DragRecorderWidget
            setTouchListener(recordButton, recorderWidget)
            return recorderWidget
        }

        fun initFloatAudioRecorder(context: Context, dataBinder: DataBinder): RecorderWidget {
//            var downX = 0F
//            var downY = 0F
            val recorderWidget: FloatRecorderWidget = RecorderWidgetFactory.getRecorderWidget(DataBinder.STYLE_FLOAT, context, dataBinder) as FloatRecorderWidget
            recorderWidget.setStatus(RecorderWidget.STATUS_IDLE)
            setTouchListener(recorderWidget.recorderView, recorderWidget)

//            recorderWidget.recorderView.setOnTouchListener({ view, motionEvent ->
//                when (motionEvent.action) {
//                    MotionEvent.ACTION_DOWN -> {
//                        downX = motionEvent.rawX
//                        downY = motionEvent.rawY
//                        return@setOnTouchListener true
//                    }
//                    MotionEvent.ACTION_MOVE -> {
//                        view.animate().translationX(motionEvent.rawX - downX)
//                                .translationY(motionEvent.rawY - downY)
//                                .setDuration(0)
//                                .start()
//                        return@setOnTouchListener true
//                    }
//                }
//                return@setOnTouchListener false
//            })
            return recorderWidget
        }

        private fun setTouchListener(recordButton: View, recorderWidget: RecorderWidget?) {
            var isCanceled = false
            var downY = 0F
            recordButton.setOnTouchListener({ view, motionEvent ->
                when (motionEvent.action) {
                    MotionEvent.ACTION_DOWN -> {
                        //开始录音
                        recorderWidget?.setStatus(RecorderWidget.STATUS_START)
                        downY = motionEvent.y
                        return@setOnTouchListener true
                    }
                    MotionEvent.ACTION_MOVE -> {
                        val moveY = motionEvent.y
                        if (downY - moveY > 100) {
                            recorderWidget?.setStatus(RecorderWidget.STATUS_DRAG)
                            isCanceled = true
                        } else {
                            recorderWidget?.setStatus(RecorderWidget.STATUS_RESUME)
                            isCanceled = false
                        }
                        return@setOnTouchListener true
                    }
                    MotionEvent.ACTION_UP -> {
                        //停止录音
                        recordButton.performClick()
                        if (isCanceled) {
                            //取消录音
                            recorderWidget?.setStatus(RecorderWidget.STATUS_CANCEL)
                            isCanceled = false
                        } else {
                            //完成录音
                            recorderWidget?.setStatus(RecorderWidget.STATUS_FINISH)
                        }
                        return@setOnTouchListener true
                    }
                }
                false
            })

        }

    }

}